### ISSUE TITLE
Proof of concept

### ISSUE THREAD

**Issue by guzart** [sveltejs/svelte-loader#1]
The svelte compiler API is very simple here's something I wrote to start playing with svelte

```javascript
const path = require('path');
const svelte = require('svelte');
const debug = require('debug')('svelte-loader');

function getFormat(target) {
  if (target === 'web') {
    return 'cjs'; // amd umd
  }

  return 'cjs';
}

module.exports = function svelteLoader(source) {
  if (this.cacheable) this.cacheable();

  const callback = this.callback;
  const filename = this.resourcePath;
  const name = path.basename(filename).replace(path.extname(filename), '');

  const format = getFormat(this.target);
  const result = svelte.compile(source, {
    // filename,
    format,
    name,
    onerror: function(err) {
      this.emitError(err.string);
    }.bind(this),
    onwarning: function(warn) {
      this.emitWarn(warn.string);
    }.bind(this),
  });

  callback(null, result.code, result.map);
};
```

I found a possible bug when trying to get sourcemaps to work using the `filename` option 

https://github.com/sveltejs/svelte/issues/43

**Comment by TheLarkInn** [comment]
This is a great start. Since I'm not familiar with the compiler API:

Is there compiler options that someone would want to pass through from webpack config loader options?

**Comment by Rich-Harris** [comment]
The compiler options are documented [here](https://github.com/sveltejs/svelte#api) – looks like they're basically all accounted for above (presumably `amd` options aren't applicable in this context?):

```js
const result = svelte.compile(source, {
  // filename,
  format,
  name,
  onerror: function(err) {
    this.emitError(err.string);
  }.bind(this),
  onwarning: function(warn) {
    this.emitWarn(warn.string);
  }.bind(this),
});
```

`filename` is commented out because I need to fix something in MagicString (that option isn't actually used anywhere *yet*, so it's not a problem). Does Webpack handle `import` natively? (Or is that only Webpack 2?) If so, it might be better to use `format: 'es'` rather than `format: 'cjs'`.

**Comment by guzart** [comment]
Almost all API options are accounted for, no need for query parsing yet. I think the most complicated part of the loader will be cross-webpack versions compatibility. Only tested above code in webpack@2.1.0-beta.27

The svelte `amd` and `umd` options are probably needed for [Webpack 1](https://webpack.github.io/docs/configuration.html#output-librarytarget) when targeting `amd` or `umd` respectively, I looked into the Webpack 2 docs, but I only found the `web` and `node` targets.

**Comment by TheLarkInn** [comment]
Yes webpack 2 will coven imports and exports naïvely without babel so an 'es' target for svelte is a good option. We can support both via the loaderContext's (`this`) `version` property https://webpack.js.org/api/loaders/#version or by reading the webpack package.json and storing it in a runtime referenced in the loader fn.

**Comment by TheLarkInn** [comment]
Or even better just a plugin that reads the package.json andsets a custom loaderContext property https://webpack.js.org/api/loaders/#custom-loadercontext-properties

**Comment by TheLarkInn** [comment]
Another thing to add which would be a pretty neat feature is if the component templates have any `<style></style>` tags for inline scoped styles, I believe we could mark any background image or URL imports or custom pre-processor using `addDependency` this would give you the traditional benefits such as using file-loader or sass-loader or url-loader, etc. That way if a component is lazy loaded, the referenced images and marked dependencies are also lazy loaded with the component.

**Comment by guzart** [comment]
I tried using `format: 'es'` but then I would have to explicitly call `.default` on the imported module.

Example:
```javascript
import App from './App.html';
new App.default({ target: document.getElementById('main') }); 
```

Am guessing is related to Svelte compiler generating the following export expression: `exports["default"] = App;`, should it be `export default App;`?

**Comment by Rich-Harris** [comment]
Hmm. In `es` mode, Svelte will output `export default SvelteComponent;` (or whatever is specified in `options.name`). In `cjs` mode it will output `module.exports = SvelteComponent;`. So you should never have a situation where you have to do `App.default`. Is Webpack (or a Webpack plugin) doing something to the module in between?

**Comment by guzart** [comment]
Confirm that when using `format: 'es'` the svelte compiler returns `export default App;`, will look deeply into my webpack configuration

**Comment by nikku** [comment]
I've put together a [basic implementation](https://github.com/nikku/svelte-loader) with tests and stuff and I'd be happy to contribute it.

**Comment by Rich-Harris** [comment]
@nikku or @guzart – sorry for the delay on my end working through other stuff, would be good to get this out in an 'official' capacity. @nikku since you have a repo up and running with tests etc already, how would you feel about transferring it to this organization (but keeping admin rights etc, of course)? If that makes sense, I can remove this repo to make room for it.

**Comment by nikku** [comment]
I'd be happy to do this.

**Comment by nikku** [comment]
Just add an issue to my repository and I initiate the transfer.
